lib/fluent/plugin/out_google_cloud.rb

```diff
@@ -665,14 +688,141 @@ def write(chunk)
+                @log.warn "Dropping #{partial_error_count} log message(s)",
+                          error_code: "google.rpc.Code[#{error_code}]",
+                          error: error_message
+              end
+            end
           end
         end
       end
     end
 
+    # Extract a map of error details from an potentially partially successful
+    # request. The keys in this map are stringified indexes of the failed log
+    # entries in the original json payload, and the values are the error code
+    # and error message.
+    #
+    # A sample error.body looks like:
+    # {
+    #   "error": {
+    #     "code": 403,
+    #     "message": "User not authorized.",
+    #     "status": "PERMISSION_DENIED",
+    #     "details": [
+    #       {
+    #         "@type": "type.googleapis.com/google.logging.v2.WriteLogEntriesPar
+    #           tialErrors",
+    #         "logEntryErrors": {
+    #           "0": {
+    #             "code": 7,
+    #             "message": "User not authorized."
+    #           },
+    #           "1": {
+    #             "code": 3,
+    #             "message": "Log name contains illegal character :"
+    #           },
+    #           "3": {
+    #             "code": 3,
+    #             "message": "Log name contains illegal character :"
+    #           }
+    #         }
+    #       },
+    #       {
+    #         "@type": "type.googleapis.com/google.rpc.DebugInfo",
+    #         "detail": ...
+    #       }
+    #     ]
+    #   }
+    # }
+    #
+    # The root level "code", "message", and "status" simply match the root
+    # cause of the first failed log entry. For example, if we switch the order
+    # of the log entries, then we would get:
+    # {
+    #    "error" : {
+    #       "code" : 400,
+    #       "message" : "Log name contains illegal character :",
+    #       "status" : "INVALID_ARGUMENT",
+    #       "details": ...
+    #    }
+    # }
+    # We will ignore it anyway and look at the details instead which includes
+    # info for all failed log entries.
+    #
+    # In this example, the logEntryErrors we are extracting should be:
+    # {
+    #   "0": {
+    #     "code": 7,
+    #     "message": "User not authorized."
+    #   },
+    #   "1": {
+    #     "code": 3,
+    #     "message": "Log name contains illegal character :"
+    #   },
+    #   "3": {
+    #     "code": 3,
+    #     "message": "Log name contains illegal character :"
+    #   }
+    # }
+    def extract_log_entry_errors(error)
+      error_details = JSON.parse(error.body)['error']['details']
+      partial_errors = error_details.detect do |error_detail|
+        error_detail['@type'] == PARTIAL_ERROR_FIELD
+      end
+      partial_errors['logEntryErrors']
+    rescue JSON::ParserError => e
+      @log.warn 'Failed to extract log entry errors from the error details' \
+                " because of JSON parsing error: #{error.body}.", error: e
+      {}
+    rescue => e
```

Comment:
Why do you need the second `rescue`?

Reply:
Because the `error`, `details` or `logEntryErrors` fields might not exist. Or there is no error details with `@type` equal to `PARTIAL_ERROR_FIELD`.

Reply:
In those cases, you'd get a `NoMethodError` on `NilClass` (or a `TypeError` if the parsed error is not a Hash), which is not going to be very useful to the user. But we could prevent those by returning an empty hash (`{}`) in all those cases, e.g.:
```ruby
    def ensure_hash(value)
      if value.is_a?(Hash) then value else {} end
    end

    def extract_log_entry_errors(error)
      error_body = ensure_hash(JSON.parse(error.body))
      error = ensure_hash(error_body['error'])
      error_details = ensure_hash(error['details'])
      partial_errors = error_details.detect do |error_detail|
        error_detail['@type'] == PARTIAL_ERROR_FIELD
      end
      ensure_hash(partial_errors['logEntryErrors'])
    rescue JSON::ParserError => e
      @log.warn 'Failed to extract log entry errors from the error details' \
                " because of JSON parsing error: #{error.body}.", error: e
      {}
    end
```
This will also make sure that the value going into `construct_error_details_map` is a Hash.

Reply:
Changed.